### PR TITLE
Promote a canary from NPM to be a stable release

### PR DESCRIPTION
**Pull request by bvaughn**
#### Issue #14201. Builds on top of #14244 ([view the delta](https://github.com/bvaughn/react/compare/automated-release-scripts-part-2...bvaughn:automated-release-scripts-part-3?expand=1)).

New script to prepare a canary release to be promoted to stable. This script checks out canary packages from NPM, prompts for new stable versions, and updates the versions (including dependencies) in the checked-out packages.

**[Learn more about these new release scripts here.](https://github.com/bvaughn/react/blob/automated-release-scripts-part-4/scripts/release/README.md#react-release-scripts)**

```bash
$ scripts/release/prepare-stable.js --version=0.0.0-5bf84d292
```

prepare-canary-2

### CODE REVIEW

File: scripts/release/create-build-commands/npm-pack-and-unpack.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    });
+  }
+
+  await exec('mkdir build');
+  await exec('mkdir build/node_modules');
+  await exec(
+    `cp -r ${tempDirectory}/build/node_modules/*.tgz ${cwd}/build/node_modules/`
+  );
+
+  // Unpack packages and parepare to publish.
```
Comment (matpag) [inline scripts/release/create-build-commands/npm-pack-and-unpack.js]: typo on "parepare"

File: scripts/release/create-canary.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env node
+
+'use strict';
+
+const {tmpdir} = require('os');
+const {join} = require('path');
+const {getBuildInfo, handleError} = require('./utils');
+
+// This local build script exists for special case, manual builds.
+// The typical suggesgted release process is to create a canary from a CI artifact.
```
Comment (matpag) [inline scripts/release/create-canary.js]: typo on "suggesgted"

File: scripts/release/prepare-stable-commands/update-stable-version-numbers.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+          // To do this, we look at both the local version of the scheduler (e.g. 0.11.0),
+          // and the dependency constraint in the local version of react-dom (e.g. scheduler@^0.11.0).
+          const sourceDependencyVersion = sourcePackageJSONs.get(dependencyName)
+            .version;
+          const sourceDependencyConstraint = sourceDependencies[dependencyName];
+
+          // If the source dependency's version and the constraint match,
+          // we will need to update the constraint to point at the dependency's new release version,
+          // (e.g. scheduler@^0.11.0 becomes scheduler@^0.12.0 when we release scheduler 0.12.0).
+          // Othewise we leave the constraint alone (e.g. react@^16.0.0 doesn't change between releases).
```
Comment (matpag) [inline scripts/release/prepare-stable-commands/update-stable-version-numbers.js]: typo on "Othewise"

File: scripts/release/prepare-stable-commands/update-stable-version-numbers.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+          }
+        }
+      }
+    }
+  };
+
+  // Update all package JSON versions and their dependencies/peerDependencies.
+  // This must be done in a way that respects semver constraints (e.g. 16.7.0, ^16.7.0, ^16.0.0).
+  // To do this, we use the dependencies defined in the source package JSONs,
+  // because the canary dependencies have already been falttened to an exact match (e.g. 0.0.0-ddaf2b07c).
```
Comment (matpag) [inline scripts/release/prepare-stable-commands/update-stable-version-numbers.js]: typo on "falttened"

**Comment by bvaughn** [comment]
Thanks. I can't type. I'll fix those.

**Comment by bvaughn** [comment]
Typo fixes included in next PR, #14260

**Comment by threepointone** [comment]
tried running this, what do you suppose I'm doing wrong? 
screenshot 2018-11-23 at 20 35 14

**Comment by bvaughn** [comment]
I'm guessing you didn't actually publish the canary you're trying to promote to stable 😄

You can either run with a `--local` flag to skip the NPM checkout (and just use whatever already happens to be in `dist`) or you can publish first, and then promote (the way the script is actually meant to be used).

**Comment by bvaughn** [comment]
Already merged via #14260

**Comment by threepointone** [comment]
aha that's what it was.

**Pull request closed without merging**